2/answers.txt

```diff
@@ -0,0 +1,231 @@
+Each thread has its own stack, but share the heap for the process entire, meaning there is inherently some shared nature of memory within the heap, but the stack itself is isolated.
+
+Q22: What is CPU IPC and frequency? Why is CPU speed more important for single threaded applications than multi-threaded applications?
+
+Instructions per clock or cycle. CPU speed is more important for a single threaded application because that application will only be functioning on a single core, as a result the speed of that core
+will dictate the speed of that program. With multithreaded applications, each thread can be split between cores in a type of load balancing, making the overall impact of core speed less for that application.
+
+Q23: What is the difference between concurrent and parallel execution? What does 'thread safety' mean?
+
+Concurrent execution is when multiple processes are being worked upon at the same time, but not executed at the same time, whereas parallel execution is where 2 or more processes are being executed simultaneously.
+An application can be any number of the above, i.e. concurrent but not parallel, parallel but not concurrent, neither and both. Thread safety applies specifically to multi threaded code. 
+
+Thread safety is about ensuring that data structure manipulated by multiple threads affect the data in the intended way, without causing unexpected changes or damage as a result of multiple entry points to the data structure.
+Typically these are done by implementing race conditions, however you can achieve thread safety without race conditions by implementing, Re-Entrancy, thread-local storage, immutable objects, mutual exclusion and atomic operations.
+
+Q24: If a CPU has 4 cores, how many processes can it execute concurrently?
+
+A CPU with 4 cores can run as many processes "Concurrently" as it has memory to store the not currently executing processes.
+
+Q25: Explain the difference between synchronous and asynchronous code. When would you prefer one technique over the other?
+
+Synchronous code has to be sequentially completed before the next set of instructions can be processed, asynchronous code allow an instruction set to be hung and another started and then returned back to the previous and so on.
+
+You would use asynchronous techniques where you can run two tasks simultaneously, usually when they do not share dependencies and are independant, and you would use synchronous techniques, when you need a task to complete before proceeding.
+
+Q26: Can a linux web-server operate acceptably with 1CPU and 512MB RAM? Can a Windows server do the same?
+
+Although Linux can run on very little RAM as little as 8MB, the act of serving as a webserver typically requires as much as 2GB RAM to functional minimally, depending of course of the load.
+The same is true of windows, except that the native cost of running windows is higher than linux and typically requires as much as 512MB of RAM to function at all.
```

Comment:
Real world numbers are a quite different from my experience.

You could happily run a C/C++/rust web server on a 512MB ram linux VM and handle thousands of concurrent requests.
Garbage collected languages need more breathing room, but it's still more than doable if your database resides on a different server.

A 1GB ram 1vcpu Linux VM can fit nginx, wordpress (which is a resource hog), an sql database, all running inside docker. It would work acceptably for a hundred concurrent reqs/s or so. Cold start in 30 seconds.

You can't even install Docker on a windows VM with less than 4gb memory.
You should try boot windows server in hyper-v with 512mb/1gb memory and do something non-trivial. it will take forever to boot and page itself to death when you try to open the management UI.